src/Common/ISlotControl.h

```diff
@@ -0,0 +1,76 @@
+#pragma once
+
+#include <limits>
+#include <memory>
+#include <base/types.h>
+#include <boost/core/noncopyable.hpp>
+
+
+namespace DB
+{
+
+// Interfaces for abstract "slot" allocation and control.
+// Slot is a virtual entity existing in a limited amount (CPUs or memory chunks, etc).
+//
+// Every slot can be in one of the following states:
+//  * free: slot is available to be allocated.
+//  * allocated: slot is allocated to a specific ISlotAllocation.
+//
+// Allocated slots can be considered as:
+//  * granted: allocated, but not yet acquired.
+//  * acquired: acquired using IAcquiredSlot.
```

Comment:
```suggestion
// Allocated slots can be in one of the following states:
//  * granted: allocated, but not yet acquired.
//  * acquired: a granted slot becomes acquired by using IAcquiredSlot.
```

I think with this it is obvious that once a slot gets acquired it "stops being granted". I think this was a bit of confusing me (after reading for the second time it is really obvious though), because it was not clear whether a slot can be "labelled" as granted and acquired at the same time. I haven't seen "granted" and "acquired" to be used this way, but I honestly don't know better wording for it.